File: cpp/tools/make_replay_save.py
#!/usr/bin/env python3
"""Generate a game-compatible save JSON for replay import.

Compatible with js/SaveSystem.importFromFile() format.
"""

from __future__ import annotations

import argparse
import json
import time
from pathlib import Path


def make_level_key(level_id: str, color: int) -> str:
    return f"{level_id}_c{color}" if color > 0 else level_id


def main() -> int:
    p = argparse.ArgumentParser(description="Generate importable replay save JSON")
    p.add_argument("--level-id", required=True, help="Level ID used by game (e.g. 1, 61, custom_1-1, gen_0)")
    p.add_argument("--replay", required=True, help="Replay string, only U/D/L/R")
    p.add_argument("--steps", type=int, default=None, help="Best steps (default: len(replay))")
    p.add_argument("--color", type=int, default=0, help="Color override for official variants (1/2/3/4/6)")
    p.add_argument("--output", default="banmen_save_import.json", help="Output json path")
    args = p.parse_args()

    replay = args.replay.strip().upper()
    if any(ch not in "UDLR" for ch in replay):
        raise SystemExit("replay must contain only U/D/L/R")

    steps = args.steps if args.steps is not None else len(replay)
    level_key = make_level_key(args.level_id, args.color)

    payload = {
        "version": 1,
        "timestamp": int(time.time() * 1000),
        "game": "BanmenHTML5",
        "content": {
            "levels": {
                level_key: {
                    "cleared": True,
                    "bestSteps": steps,
                    "replay": replay,
                }
            }
        },
    }

    out = Path(args.output)
    out.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Wrote {out}")
    print(f"level_key={level_key}")
    print(f"steps={steps}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: cpp/tools/make_replay_save.cpp
#include <algorithm>
#include <chrono>
#include <fstream>
#include <iostream>
#include <string>

static void usage() {
    std::cout
        << "Usage: make_replay_save --level-id <id> --replay <UDLR...> [--steps N] [--color N] [--output path]\n";
}

static bool isReplayValid(const std::string& replay) {
    return std::all_of(replay.begin(), replay.end(), [](char c) {
        return c == 'U' || c == 'D' || c == 'L' || c == 'R';
    });
}

int main(int argc, char** argv) {
    std::string levelId;
    std::string replay;
    std::string output = "banmen_save_import.json";
    int color = 0;
    int steps = -1;

    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        auto needValue = [&](const std::string& name) -> const char* {
            if (i + 1 >= argc) {
                std::cerr << "Missing value for " << name << "\n";
                std::exit(2);
            }
            return argv[++i];
        };

        if (a == "--level-id") levelId = needValue(a);
        else if (a == "--replay") replay = needValue(a);
        else if (a == "--steps") steps = std::stoi(needValue(a));
        else if (a == "--color") color = std::stoi(needValue(a));
        else if (a == "--output") output = needValue(a);
        else if (a == "--help" || a == "-h") {
            usage();
            return 0;
        } else {
            std::cerr << "Unknown arg: " << a << "\n";
            usage();
            return 2;
        }
    }

    if (levelId.empty() || replay.empty()) {
        usage();
        return 2;
    }

    std::transform(replay.begin(), replay.end(), replay.begin(), [](unsigned char c) {
        return static_cast<char>(std::toupper(c));
    });

    if (!isReplayValid(replay)) {
        std::cerr << "replay must contain only U/D/L/R\n";
        return 2;
    }

    if (steps < 0) steps = static_cast<int>(replay.size());
    const std::string levelKey = color > 0 ? (levelId + "_c" + std::to_string(color)) : levelId;

    const auto now = std::chrono::time_point_cast<std::chrono::milliseconds>(
                         std::chrono::system_clock::now())
                         .time_since_epoch()
                         .count();

    std::ofstream ofs(output, std::ios::binary);
    if (!ofs) {
        std::cerr << "Cannot write output: " << output << "\n";
        return 1;
    }

    ofs << "{\n"
        << "  \"version\": 1,\n"
        << "  \"timestamp\": " << now << ",\n"
        << "  \"game\": \"BanmenHTML5\",\n"
        << "  \"content\": {\n"
        << "    \"levels\": {\n"
        << "      \"" << levelKey << "\": {\n"
        << "        \"cleared\": true,\n"
        << "        \"bestSteps\": " << steps << ",\n"
        << "        \"replay\": \"" << replay << "\"\n"
        << "      }\n"
        << "    }\n"
        << "  }\n"
        << "}\n";

    std::cout << "Wrote " << output << "\n";
    std::cout << "level_key=" << levelKey << "\n";
    std::cout << "steps=" << steps << "\n";
    return 0;
}
